// @mongosh/cli-repl/lib/run.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const index_1 = require("./index");
const service_provider_server_1 = require("@mongosh/service-provider-server");
const history_1 = require("@mongosh/history");
const module_1 = require("module");
(async () => {
    if (process.env.MONGOSH_RUN_NODE_SCRIPT) {
        if (process.execPath !== process.argv[1]) {
            process.argv.splice(1, 1);
        }
        module_1.runMain(process.argv[1]);
        return;
    }
    let repl;
    try {
        const options = index_1.parseCliArgs(process.argv);
        const { version } = require('../package.json');
        if (options.help) {
            console.log(index_1.USAGE);
        }
        else if (options.version) {
            console.log(version);
        }
        else if (options.buildInfo) {
            console.log(JSON.stringify(index_1.buildInfo(), null, '  '));
        }
        else if (options.smokeTests) {
            const smokeTestServer = process.env.MONGOSH_SMOKE_TEST_SERVER;
            if (process.execPath === process.argv[1]) {
                await index_1.runSmokeTests(smokeTestServer, process.execPath);
            }
            else {
                await index_1.runSmokeTests(smokeTestServer, process.execPath, process.argv[1]);
            }
        }
        else {
            let mongocryptdSpawnPaths = [['mongocryptd']];
            if (process.execPath === process.argv[1]) {
                process.removeAllListeners('warning');
                mongocryptdSpawnPaths = await index_1.getMongocryptdPaths();
            }
            if (process.env.CLEAR_SIGINT_LISTENERS) {
                process.removeAllListeners('SIGINT');
            }
            const driverOptions = index_1.mapCliToDriver(options);
            const driverUri = service_provider_server_1.generateUri(options);
            const title = `mongosh ${history_1.redactURICredentials(driverUri)}`;
            process.title = title;
            setTerminalWindowTitle(title);
            const appName = `mongosh ${version}`;
            const shellHomePaths = index_1.getStoragePaths();
            repl = new index_1.CliRepl({
                shellCliOptions: {
                    ...options,
                },
                mongocryptdSpawnPaths,
                input: process.stdin,
                output: process.stdout,
                onExit: process.exit,
                shellHomePaths: shellHomePaths
            });
            await repl.start(driverUri, { appName, ...driverOptions });
        }
    }
    catch (e) {
        console.error(`${e.name}: ${e.message}`);
        if (repl !== undefined) {
            repl.bus.emit('mongosh:error', e);
        }
        process.exit(1);
    }
})();
function setTerminalWindowTitle(title) {
    var _a;
    if (!process.stdout.isTTY) {
        return;
    }
    const term = (_a = process.env.TERM) !== null && _a !== void 0 ? _a : '';
    if (/^(linux|xterm|rxvt)/.test(term)) {
        process.stdout.write(`\u001b]0;${title}\u0007`);
    }
    else if (/^screen/.test(term)) {
        process.stdout.write(`\u001bk${title}\u001b\\`);
    }
}
